### PR TITLE
product-type

### PR DESCRIPTION
**Pull request by mabbus**

### CODE REVIEW

File: app/Http/Controllers/Web/EmployeesDealsController.php [hunk 1]
```diff
@@ -42,6 +43,10 @@ public function index()
             ['company_id', '=', data_get($company, 'id')]
         ]);
 
+        if($request->query('product_type')) {
```
Comment (lexod) [inline app/Http/Controllers/Web/EmployeesDealsController.php:46]: should come from request->validated()

File: app/Http/Requests/Deal/CreateDeal.php [hunk 2]
```diff
@@ -30,12 +30,13 @@ public function rules()
...
-            'terms' => 'required|integer|min:1',
-            'term_type' => 'in:0,1,2,3,4,5',
+            'terms' => 'integer|min:1',
+            'term_type' => 'nullable|in:0,1,2,3,4,5,6',
             'amount' => 'required|numeric|min:1|regex:/^\d+(\.\d{1,2})?$/',
             'rate' => 'required|numeric|between:1.000,2.000',
             'buy_rate' => 'required|numeric|between:1.000,2.000',
-            'closing_fee' => 'required|numeric|between:0,99.999',
+            'closing_fee' => 'numeric|between:0,99.999',
+            'product_type_id' => 'required|integer',
```
Comment (lexod) [inline app/Http/Requests/Deal/CreateDeal.php:39]: add exists check

File: app/Http/Tasks/ProductType/GetProductTypes.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+class GetProductTypes implements TaskInterface
+{
+    public function __construct()
+    {
+    }
+
+    public function run()
+    {
+        return ProductType::all();
```
Comment (lexod) [inline app/Http/Tasks/ProductType/GetProductTypes.php:16]: this place is a potential performance issue
1. select only those columns you need
1. restrict it 5 items
1. you don't need a task for such a simple action

File: app/Http/Tasks/Deal/AdminGetDeals.php [hunk 4]
```diff
@@ -37,6 +37,10 @@ public function run()
 
         $deals = Deal::with('business');
 
+        if(Arr::has($data, 'product_type')) {
```
Comment (lexod) [inline app/Http/Tasks/Deal/AdminGetDeals.php:40]:
```php
$deals->when(Arr::has($data, 'product_type'), function($query) use ($data) {return $query->where('product_type_id', $data['product_type']);});
```

File: app/Http/Traits/ValidationTrait.php [hunk 5]
```diff
@@ -80,6 +80,7 @@ public function getPaginationRules()
         return [
             'offset' => 'nullable|integer|min:0',
             'limit' => 'nullable|integer|min:1|max:100',
+            'product_type' => 'nullable|integer|min:1'
```
Comment (lexod) [inline app/Http/Traits/ValidationTrait.php:83]: not meant to be here,
needs to be a part of each request.

File: database/migrations/2020_10_26_132118_create_deal_product_type_table.php [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('deal_product_types', function (Blueprint $table) {
+            $table->increments('id');
+            $table->string('name', 200)->nullable();
```
Comment (lexod) [inline database/migrations/2020_10_26_132118_create_deal_product_type_table.php:18]: what's the point to have it nullable?

**Review by lexod — CHANGES_REQUESTED** [review]

File: app/Http/Tasks/ProductType/GetProductTypes.php (hunk 3 above)
Comment (mabbus) [reply app/Http/Tasks/ProductType/GetProductTypes.php:16]: we eventually need to filter by company_id and stuff

File: app/Http/Controllers/Web/EmployeesDealsController.php [hunk 7]
```diff
@@ -22,26 +23,38 @@
...
         ]);
 
+        if(Arr::get($data,'product_type')) {
+            $deals->where('product_type_id', $data['product_type']);
+        }
+
+        if(Arr::get($data, 'workflow_status')) {
+                $deals->where('status', $data['workflow_status']);
+        }
+
```
Comment (lexod) [inline app/Http/Controllers/Web/EmployeesDealsController.php:57]: user `->when()`

File: app/Http/Controllers/Web/EmployeesDealsController.php [hunk 8]
```diff
@@ -59,17 +72,30 @@ public function index()
             'progress' => '',
         ];
 
-        $statuses = WorkflowStatus::all();
+        $statuses = WorkflowStatus::withCount(['deals' => function ($q) use ($data) {
+            if(Arr::get($data, 'product_type')) {
```
Comment (lexod) [inline app/Http/Controllers/Web/EmployeesDealsController.php:76]: user `->when()`

File: app/Http/Tasks/Deal/AdminGetDeals.php [hunk 9]
```diff
@@ -37,6 +37,14 @@ public function run()
...
         $deals = Deal::with('business');
 
+        if(Arr::get($data, 'product_type')) {
+            $deals->where('product_type_id', $data['product_type']);
+        }
+
+        if(Arr::get($data, 'workflow_status')) {
+            $deals->where('status', $data['workflow_status']);
+        }
+
```
Comment (lexod) [inline app/Http/Tasks/Deal/AdminGetDeals.php:47]: user `->when()`

**Review by lexod — CHANGES_REQUESTED** [review]

**Review by lexod — APPROVED** [review]

**Pull request merged**